src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -403,105 +403,124 @@ exprt string_constraint_generatort::add_axioms_for_correct_number_format(
-  exprt starts_with_digit=
-    not_exprt(or_exprt(starts_with_minus, starts_with_plus));
+  /// We experimented with making the max_string_length depend on the base, but
+  /// this did not make any difference to the speed of execution.
+  const std::size_t max_string_length=to_bitvector_type(type).get_width()+1;
 
   /// TODO: we should throw an exception when this does not hold:
-  const std::size_t max_string_length=40;
+  /// Note that the only thing stopping us from taking longer strings with many
+  /// leading zeros is the axioms for correct number format
   const exprt &correct=add_axioms_for_correct_number_format(
     str, radix, max_string_length);
   axioms.push_back(correct);
 
-  /// TODO(OJones): size should depend on the radix
-  /// TODO(OJones): we should deal with overflow properly
-  for(std::size_t size=1; size<=max_string_length; size++)
-  {
-    exprt sum=from_integer(0, type);
-    exprt first_value=get_numeric_value_from_character(chr, char_type, type);
-    equal_exprt premise=str.axiom_for_has_length(size);
+  /// TODO(OJones): Fix the below algorithm to make it work for min_int. There
+  /// are two problems. (1) Because we build i as positive and then negate it if
+  /// the first character is '-', we hit overflow for min_int because
+  /// |min_int| > max_int. (2) radix^63 overflows. I think we'll have to
+  /// special-case it.
 
-    for(std::size_t j=1; j<size; j++)
-    {
-      mult_exprt radix_sum(sum, radix);
-      if(j>=max_string_length-1)
-      {
-        // We have to be careful about overflows
-        div_exprt div(sum, radix);
-        implies_exprt no_overflow(premise, (equal_exprt(div, sum)));
-        axioms.push_back(no_overflow);
-      }
+  axioms.push_back(binary_relation_exprt(x, ID_ge, zero_expr));
 
-      sum=plus_exprt_with_overflow_check(
-        radix_sum,
-        get_numeric_value_from_character(str[j], char_type, type));
+  exprt radix_to_power_of_i;
+  exprt radix_to_power_of_i_minus_one;
+  exprt no_overflow;
 
-      mult_exprt first(first_value, radix);
-      if(j>=max_string_length-1)
-      {
-        // We have to be careful about overflows
-        div_exprt div_first(first, radix);
-        implies_exprt no_overflow_first(
-          and_exprt(starts_with_digit, premise),
-          equal_exprt(div_first, first_value));
-        axioms.push_back(no_overflow_first);
-      }
-      first_value=first;
-    }
+  for(std::size_t i=0; i<max_string_length; ++i)
+  {
+    /// We are counting backwards from the end of the string, i.e. i refers
+    /// to str[j] where j=str.length()-i-1
+    const constant_exprt i_expr=from_integer(i, index_type);
+    const minus_exprt j(
+      minus_exprt(str.length(), i_expr), from_integer(1, index_type));
 
-    // If the length is `size`, we add axioms:
-    // a1 : starts_with_digit => i=sum+first_value
-    // a2 : starts_with_plus => i=sum
-    // a3 : starts_with_minus => i=-sum
+    if(i==0)
+    {
+      no_overflow=true_exprt();
```

Comment:
On closer inspection, that variable doesn't need to be declared outside of the branch in which it is initialised.